tox_extension_messages.c

```diff
@@ -157,100 +194,236 @@ bool parse_messages_packet(uint8_t const *data, size_t size,
-		}
+	if (extension->cb) {
+		extension->cb(friend_id, incoming_message->message,
+			      incoming_message->size, extension->userdata);
+	}
 
-		uint8_t data[9];
-		data[0] = MESSAGE_RECEIVED;
-		toxext_write_to_buf(parsed_packet.receipt_id, data + 1, 8);
-		toxext_packet_append(response_packet,
-				     ext_message_ids->extension_handle, data,
-				     9);
+	uint8_t data[9];
+	data[0] = MESSAGE_RECEIVED;
+	toxext_write_to_buf(parsed_packet->receipt_id, data + 1, 8);
+	toxext_segment_append(response_packet, extension->extension_handle,
+			      data, 9);
+
+	clear_incoming_message(incoming_message);
+}
+
+void tox_extension_messages_handle_message_part(
+	struct ToxExtensionMessages *extension, uint32_t friend_id,
+	struct MessagesPacket *parsed_packet,
+	struct IncomingMessage *incoming_message)
+{
+	struct FriendSendingInvalidSize *friend_dropping =
+		get_friend_dropping(extension, friend_id);
+	if (friend_dropping->sending_invalid) {
+		/* FIXME: We should probably tell the sender that we dropped a message here */
+		clear_incoming_message(incoming_message);
+		return;
+	}
+	tox_extension_copy_in_message_data(parsed_packet, incoming_message);
+}
+
+// Here they confirm that they are indeed sending us messages. We don't call the
+// negotation callback when we've both negotiated that we have the extension
+// because at that point they don't know we can accept the message ids. We wait for
+// an enable flag from the other side to indicate that they are now embedding message ids.
+static void tox_extension_messages_recv(
+	struct ToxExtExtension *extension, uint32_t friend_id, void const *data,
+	size_t size, void *userdata, struct ToxExtPacketList *response_packet)
+{
+	(void)extension;
+	struct ToxExtensionMessages *ext_messages = userdata;
+	struct IncomingMessage *incoming_message =
+		get_incoming_message(ext_messages, friend_id);
 
+	struct MessagesPacket parsed_packet;
+	if (!parse_messages_packet(data, size, &parsed_packet)) {
+		/* FIXME: We should probably tell the sender that they gave us invalid data here */
 		clear_incoming_message(incoming_message);
+		return;
+	}
+
+	switch (parsed_packet.message_type) {
+	case MESSAGE_NEGOTIATE:
+		tox_extension_messages_negotiate_size_cb(
+			ext_messages, friend_id,
+			parsed_packet.max_sending_message_size);
+		return;
+	case MESSAGE_START:
+		tox_extension_messages_handle_message_start(ext_messages,
+							    friend_id,
+							    &parsed_packet,
+							    incoming_message);
+		return;
+	case MESSAGE_PART: {
+		tox_extension_messages_handle_message_part(ext_messages,
+							   friend_id,
+							   &parsed_packet,
+							   incoming_message);
+		return;
+	}
+	case MESSAGE_FINISH:
+		tox_extension_messages_handle_message_finish(
+			ext_messages, friend_id, &parsed_packet,
+			incoming_message, response_packet);
+		return;
+	case MESSAGE_RECEIVED:
+		ext_messages->receipt_cb(friend_id, parsed_packet.receipt_id,
+					 ext_messages->userdata);
+		return;
 	}
+	assert(false);
```

Comment:
Seems kinda messed up to assert if a packet has some unexpected data in it.

Reply:
Definitely, bug. Removed.